posthog/decorators.py

```diff
@@ -25,37 +34,45 @@ class CacheType(str, Enum):
     PATHS = "Path"
 
 
-def cached_function():
-    def parameterized_decorator(f: Callable):
-        @wraps(f)
-        def wrapper(*args, **kwargs) -> Dict[str, Union[List, datetime, bool, str]]:
-            # prepare caching params
-            request: Request = args[1]
-            team = cast(User, request.user).team
-            filter = None
-            if not team:
-                return f(*args, **kwargs)
-
-            filter = get_filter(request=request, team=team)
-            cache_key = generate_cache_key("{}_{}".format(filter.toJSON(), team.pk))
-            # return cached result if possible
-            if not should_refresh(request):
-                cached_result = get_safe_cache(cache_key)
-                if cached_result and cached_result.get("result"):
-                    return {**cached_result, "is_cached": True}
-            # call function being wrapped
-            result = f(*args, **kwargs)
-
-            # cache new data
-            if result is not None and not (isinstance(result.get("result"), dict) and result["result"].get("loading")):
+ResultPackage = Union[Dict[str, Any], List[Dict[str, Any]]]
+
+T = TypeVar("T", bound=ResultPackage)
+U = TypeVar("U", bound=GenericViewSet)
+
+
+def cached_function(f: Callable[[U, Request], T]) -> Callable[[U, Request], T]:
+    @wraps(f)
+    def wrapper(self, request) -> T:
+        # prepare caching params
+        team = cast(User, request.user).team
+        filter = None
+        if not team:
+            return f(self, request)
+
+        filter = get_filter(request=request, team=team)
+        cache_key = generate_cache_key("{}_{}".format(filter.toJSON(), team.pk))
+
+        # return cached result if possible
+        if not should_refresh(request):
+            cached_result_package = get_safe_cache(cache_key)
+            if cached_result_package and cached_result_package.get("result"):
+                cached_result_package["is_cached"] = True
+                return cached_result_package
```

Comment:
Do we know if `last_refresh` is always included in this result package for cached responses?

Reply:
We do, because we set it a few lines below.